### ISSUE TITLE
Away mode does not work as expected

### ISSUE THREAD

**Issue by cygnusb** [oischinger/ha_vicare#6]
I am not quite sure, if the current implementation of away mode is correct. It is not consistent with ViCare App at least with my device. It might also be that implementing hold mode would be correct for this (CLIMATE.SET_HOLD_MODE), which is currently not implemented.

Setting away on the ViCare App results in forcedReduced to be set as Mode. Setting home in ViCare changes the mode to forcedNormal.

Is this away mode in ViCare App supposed to be implemented by CLIMATE.SET_HOLD_MODE or CLIMATE.SET_AWAY_MODE?

**Comment by oischinger** [comment]
Holiday mode as it is right now implemented does not make much sense. I just added it to experiment further.

Viessmann expects a holiday schedule to be set as parameter for the holiday mode instead of just turning it on and off like a switch.
So the concept of a Home assistant holiday mode is just different.
We might think of either setting it into hold mode or instead, my device does not have that mode, set it to forcedReduced.
Turning off holiday mode would set it back to normal (not forcedNormal as it is now).

Instead we could also use a input value component where the user could set a date  which is then used as the end date of a holiday schedule.

Comments & Suggestions welcome!

**Comment by cygnusb** [comment]
That pull request works for me. Bad thing currently is, that I do not see a hold switch in the climate component. Most possibly this is not implemented in HA?!? But using an automation it is possible to set hold mode by issueing a:

climate.set_hold_mode on the corresponding entity in domain climate with data:
```
{
  "entity_id": "climate.vicare",
  "hold_mode": "away"
}
```

and back with:

```
{
  "entity_id": "climate.vicare",
  "hold_mode": "off"
}
```

**Comment by vandenberghev** [comment]
The whole away/hold/reduced modes and programs are a bit of a mess... In my opinion, the climate component should reflect what the ViCare app shows to avoid confusion. Not the internal HA stuff.

I'm willing to make some modifications, but I'm not sure I understand the difference between sensor.vicare_activemode and sensor.vicare_activeprogram? How do these two thing cooperate? Is activeprogram just the time-based schedule of the Viessmann thermostat?

And what about climate.set_hold_mode, climate.set_away_mode and climate.set_operation_mode? As far as I understand, HA's away mode is "permanent", whereas the hold mode is "temporary", whatever that means...

It's probably a good idea to list everything, in order to understand what maps to what.

The operating modes in the ViCare app are:
- Only hot water active
- Heating and hot water active (= runs the time-based 'comfort' schedules for heating and water)
- Constant day mode (= always on @ normal temp)
- Continuous reduced (= always on @ reduced temp)
- Off

The operating modes on my Viessmann thermometer are:
- Only hot water active
- Heating and hot water active
- Party mode (= constant day mode for X hours, reverts automatically)
- Eco mode (= continuous reduced mode for X time?)
- Off

The operating modes in HA are:
- Auto
- Heat
- Eco
- Off

Additional settings for HA:
- Away mode
- Hold mode

So the easy mapping is:

ViCare app | Viessmann thermometer | Home Assistant
---------- | --------------------- | --------------
Only hot water active | Only hot water active | Not supported
Heating and hot water active | Heating and hot water active | Auto
Constant day mode | External programming | Heat
Continuous reduced | External programming | Eco
Off | Off | Off

Now, according to the ViCare API...

ViCare API modes:
- dhw
- dhwAndHeating
- forcedNormal
- forcedReduced
- standby

ViCare API programs:
- active
- comfort
- eco
- external
- holiday
- normal
- reduced
- standby

ViCare modes is easy; there is a 1 on 1 mapping with the ViCare app:
ViCare API mode | ViCare app
--------------- | ----------
dhw | Only hot water active
dhwAndHeating | Heating and hot water active
forcedNormal | Constant day mode
forcedReduced | Continuous reduced
standby | Off

The ViCare API programs... that's convoluted mess. I have no idea how these work/map.
Might experiment later. I think the 'programs' are mostly just a collection of random stuff to hide the underlying 'modes'. So I believe we should not pay to much attention to them (not at this time, in any case).

So that leaves us with the implementation in the climate component.

Operation modes (the dropdown in the UI) should be pretty clear in the table above. Not sure what to do about 'Hot water only'; maybe the climate operation modes are extendable?

For me, the 'Away' button and HA's away mode are redundand, but they might be practical for some people. My suggestion for this button's behavior is: save the current mode, whenever 'Away' is activated switch to 'forcedReduced' mode, when 'Away' is deactivated, restore the previous mode.

What do you guys think?

**Comment by vandenberghev** [comment]
I implemented my proposition here: https://github.com/vandenberghev/ha_vicare/tree/climate_component_proposition

It works as inteded, but the UI is only updated after 30 seconds... Might require a manual UI update.

**Comment by cygnusb** [comment]
Thanks for writing everything down @oischinger . I fully agree that this needs a complete understanding how this is done and one should also keep in mind, that it should be possible to control the Viessmann thermostat in parallel on the device, the VICare App and integration in Home Assistant.
So settings changed on one control mechanism should always be reflected and visible in the other controls. So overwriting some stored program blindly on the device (which might have changed by some other control means) should not be done.

I understand this exactly that way, that activeProgram is the set time based schedule and the mode is to hold either the "normal" or "reduced" temperatures of the active running program (activeProgram). 

I think the operating modes of HA can be reflected by part of the activePrograms of the ViCare API.

Away Mode is a simple on/off toggle in HA. Hold mode is a multi state option, which is not strictly defined, but can have some options like:

  - home
  - away
  - off

but could also be extended further. Since the ViCare API has at least 3 possible options one could map Hold Mode in the following way:

|HA set_mode|ViCare API mode|
|-|-|
|home|forcedNormal|
|away|forcedReduced|
|auto|normal|
|off|dhwAndHeating|
|hwater ?|dhw ?|

Thats how I understood this and also implemented this (except the hot water only option).

And for the programs:

|HA program|ViCare API program|
|-|-|
|off|standby|
|away|holiday|
|eco|eco|
|heart|comfort|

**Comment by vandenberghev** [comment]
> So overwriting some stored program blindly on the device (which might have changed by some other control means) should not be done.

Why not? Whenever I change modes in the ViCare app, the only thing my thermometer shows is "External program". HASS reflects changed made both on the thermometer, and in the ViCare app . And the Vicare app correctly reflects the changes I make through HASS and the thermometer.

So my guess is that the ViCare app only changes **mode**, and always sets the **program** to 'external'. Which can/should(?) thus also exactly be what the HASS component does, IMO :) The HASS component reflects the state currently known in the cloud, thus it should not pose a problem to blindly overwrite on changes in HASS.

My 2 cents 😃 

P.S. I also added a (temperature only) DHW control panel in the climate component.

**Comment by oischinger** [comment]
I also understood modes to be a permanent thing, while programs may be only temporary.

E.g. if I turn on eco it will be on for a couple of hours (don't know how long exactly) and it will automatically be turned of by the next change from normal to reduced (if such a schedule is set).
It's the other way round for comfort mode.
This way the name "program" seems to be refering to it being either a temporary state or controlled by a schedule (holiday, normal, reduced).

Whereas "modes" seem to be a permanent configuration (e.g. dhw for hot water only or dhwAndHeating for hot water and heating)

That's why I like @cygnusb s implementation of the hold mode right now: It sets forceReduced and forcedNormal. It changes the state immediately (not like e.g. a holiday schedule) and it is not overidden by some Viessmann-internal programming sometime later.

@vandenberghev : Does your Vicare set the program to external? Mine only sets the program to eco or comfort. I cannot do much more with this app. The previous Vitotrol Plus app however allowed to set me both: programs and modes.
I guess the app functionality depends alot on which type of heater you have installed, maybe also which software it runs.

In the end I guess set_operation_mode should allow us to set all modes Vicare provides.
Concerning the program: My only use case for that is to have some sort of holiday mode where I set forceReduced. This is currently fullfilled with the hold mode @cygnusb has implemented.

Anyway, I really appreciate this discussion. I'm starting to get a clearer picture of the ViCare API, but also the way the climate component in HA.

**Comment by vandenberghev** [comment]
@oischinger I just retested the app <-> thermostat implementation.
If I set forcedNormal and forcedReduced through the app, the thermostat shows 'external program'.
dhwAndHeating + off show up as if they were set on the thermostat itself, so no external.

And dhw seems glitchy... If I set in on the thermostat, it shows up as dhw only in the app, which is good.
But if I set it through the app, it shows up as 'off' on the thermostat. Probably a bug.

I'm still not sure I understand the exact difference between 'hold' and 'away' mode in HA, or maybe it's because I do see a good use case for 'hold' mode? If it's confusing for us, it surely will be for an end-user. That's why my approach to the climate component is rather 'KISS'.

I actually have 3 use cases (for myself):

- Usually my thermostat is just set to dhwAndHeating, so time-scheduled.
- Or I'm sure I'll be away for some time, and it's set to forcedReduced.
- Or I'm home during a period that's not covered by the time schedule, so forcedNormal.

So I think that covers most cases for everyone?
What are some more use cases you guys use/see?

B.T.W. I don't know if you've seen the [lovelace climate component](https://i.imgur.com/D953I1v.png), but it's also KISS: you can set the temperature, and set 4 modes: off, forcedNormal, forcedReduced, dhwAndHeating. That's it. The [old 'states' UI](https://i.imgur.com/PmW3kHh.png) contains more elements, but it's also way more confusing; elements/buttons seem redundant.

P.S. I use the iOS app, and my thermometer is a Vitotrol 300.

**Comment by vandenberghev** [comment]
Oh... I just noticed the [climate services](https://www.home-assistant.io/components/climate/#climate-control-services)... Things are starting to make more sense now, I assumed `climate.set_operation_mode` could set any mode :) I've been misinterpreting this whole issue wrong the entire time haha

**Comment by oischinger** [comment]
Yes, the modes are free strings even though there are some predefined ones which one should (this is why you see a green leaf button for the eco state).

**Comment by oischinger** [comment]
Back to the original topic of this issue:
I implemented away mode now to set HOLD mode to "away" when turned on and "off" when
turned off.
This should lead to forcedReduced in away mode and reset to the original 
mode or dhwAndHeating when turned off.

It seems like duplication but since hold mode has no representation in the HA UI it is kind of nice to have a toggle switch for the away mode.

I'm closing the issue for now. Feel free to reopen it if you have further thoughts.
